web/src/labels.ts

```diff
@@ -46,3 +64,89 @@ export function getUiLabels({ labels }: UILabels): string[] {
 export function orderLabels(labels: string[]) {
   return [...labels].sort((a, b) => a.localeCompare(b))
 }
+
+// This helper function takes a template type for the resources
+// and a label accessor function
+export function resourcesHaveLabels<T>(
+  features: Features,
+  resources: T[] | undefined,
+  getLabels: (resource: T) => string[]
+): boolean {
+  // Labels on resources are ignored if feature is not enabled
+  if (!features.isEnabled(Flag.Labels)) {
+    return false
+  }
+
+  if (resources === undefined) {
+    return false
+  }
+
+  return resources.some((r) => getLabels(r).length > 0)
+}
+
+// Shared resource grouping styles
```

Comment:
yup, i think that's generally a correct read on how i approached this implementation. i extracted out the shared styling between the two views because there are a lot of material ui component style overrides. each view implementation adds additional styling that's specific to the view.

i didn't think about how reusable the components are outside of these two views because we're only using grouping in these places. it seemed like an okay trade-off to me, but i'm happy to revisit that, either to create more reusable versions of the components or organize the code in a different manner.

there is a lack of commenting and documentation about the relationship bw the components that i can see is causing friction!

Reply:
two patterns i've seen are:
1) the one we use in SidebarItemView, where we create a single component with top-level CSS classes to control variants on styling (https://github.com/tilt-dev/tilt/blob/master/web/src/SidebarItemView.tsx#L68)
2) the material-ui styling hooks, which we use in a few places: https://material-ui.com/styles/basics/#hook-api

i agree that the main goal is to figure out what's the best way to move this PR forward, and the way to do that might be to use this approach for now

Reply:
i think it would also be totally reasonable to make these two completely separate components with no shared code. i was reminded of this old advice this week: https://twitter.com/zkat__/status/1425859526616616960